pkg/test/framework/components/istio/operator.go

```diff
@@ -193,6 +199,35 @@ func deploy(ctx resource.Context, env *kube.Environment, cfg Config) (Instance,
 	return i, nil
 }
 
+func setupNamespaces(c *operatorComponent) error {
+	// gather references to configured names, grouped by configured value
```

Comment:
This is here to respect all the command line flags. They all have the same value by default. I'm not sure what the value is to setting these flags if we end up using generated names. Should there be something that says if "if value != default then create namespace without using test framework" or "use exact name"?

Reply:
I would think the behavior is they default to `""` which means we autogenerate as here. If they are set, we just use that and do not autogenerate

Reply:
We set them to "istio-system" by default, other test framework components like pilot and ingress read those values and expect them to be set. I made a change to say if they're != their default, don't autogenerate or create the namespace.

Reply:
I think this normalization should be done here instead: https://github.com/istio/istio/blob/master/pkg/test/framework/components/istio/config.go#L196. That way the values are always set and the other components can just use them as-is.

Reply:
I'm not sure what you mean. Removing the initial values for the namespace fields in settingsFromCommandLine, then checking `if ns == "" { ns = defaultSystemNamespace }`? 

Or do you mean including the namespace generation/creation inside `DefaultConfig`? That could cause problems because a lot of the callers make a call to DefaultConfig that is disconnected to their `istio.Setup` call. They'd end up creating an extra namespace where istio isn't deployed. Maybe it would be better if DefaultConfig was private and components needing the config would have to reference the istio component and call Settings?

Reply:
Something that requires less refactoring would to be allowing callers of istio.Setup to specify if they want generated namepace(s) for istio, that way calls to DefaultConfig will still return what they expect.

Reply:
Sorry, ignore my previous comment ... I misunderstood the intent here.

So `namespace.New` will always generate a namespace with a UUID. Is this the intent? I would expect that by default (i.e. everything is "") that we generated a single namespace `istio-system-1234` that is used for everything. All of our tests on CI would use this default for isolation. If we wanted to run a test against an existing system, however, we would want to specify the exact namespace(s) to use (e.g. istio-system). In that case we would need to do a `namespace.Claim` so that we don't append a UUID extension.

Reply:
So on the first call to default config we'd perform this initialization and save it for the rest of the tests in the session? I'd expect the behavior to be a new generated namespace for each Istio component. The only way to set a specific namespace is via command line flags, so you wouldn't be able to do that per-test.

Reply:
If newly generated namespaces for each component works, let's do it.

Reply:
We get limited value if we are just implementing the ability to configure these namespaces and `Claim`ing them if they're set. This doesn't get us the benefits of cleanup between tests or isolation in individual tests. Only the ability to isolate the entire test run from an istio-system namespace already on the k8s cluster. 

The challenge with using newly gen'd namespaces for each component is that many parts of the test framework utilize `DefaultConfig` to get the namespace istio is in. There are actually TODOs to change this in the pilot, mixer and galley components. We'd need to make all these look at an istio.Instance rather than the default config if each instance has the possibility of using non-default configuration. 

We could refactor to remove the dependency of defaultConfig, since all of the components that end up using it actually have a reference to an istio instance. It would be a very wide change though. I started it on [this branch](https://github.com/stevenctl/istio/commits/private-default-config). 

An alternative would be adding something like [this](https://github.com/stevenctl/istio/commit/567e0612569d7613e2cd59e2910b4076799eab90#diff-d4c4078f7ac2556f3be2263cd2e094cbR73) to allow getting the current istio config from the resource.Context. This would make the [refactor much simpler](https://github.com/stevenctl/istio/commit/e19ccadab4eda4a6dce8e9fef0e3ac426395a353).

Reply:
Just backing up a bit, I think there are 2 fundamental use cases we need to consider:

1. Running tests against a new system
2. Running tests against an existing system

For 1, we are controlling the lifecycle of the Istio system. We deploy it, we run our tests against it, and then we tear it down. We can theoretically run multiple tests at the same time, so we would always want to use dynamically generated namespaces to avoid conflict.

For 2, we already have an Istio system deployed. We might be debugging a production system, for example.  In this situation, we aren't controlling the lifecycle of the Istio deployment, so we would run with `istio.test.kube.deploy=false` and `istio.test.nocleanup`. We would also have to explicitly set the system namespaces (e.g. `istio-system`) so that the test harness will find all of the deployed resources.

WRT your point about `defaultConfig`, I think you're right. We could theoretically have multiple Istio components in a single application (I think John has tests that do this), and we can't use the same namespaces for both. I like the idea about adding the ability to look up a component from the context like you do [here](https://github.com/stevenctl/istio/commit/567e0612569d7613e2cd59e2910b4076799eab90#diff-d4c4078f7ac2556f3be2263cd2e094cbR73). We might want to change it to be by ID, however. So the calling code would look something like `istio.Get(ctx)[0].Config()`. I think we should just do this. Do you want to turn this into a PR and get this in first?